### ISSUE TITLE
LED strip with WS2811 controller

### ISSUE THREAD

**Issue by frankonski** [HomeSpan/HomeSpan#781]
Hi, I'm taking on a new endeavor, which is to get control over a newly purchased [Alitove Neon strip](https://www.amazon.ca/dp/B0CDQ2LDQH?ref=ppx_yo2ov_dt_b_product_details&th=1) .
I used your Holiday Lights sample to get started, and was hoping that, bare minimum, the _ManualControl_ special effect would get me started to have the strip LEDs all set to the same color.

After realizing that the strip supposedly had no white LED (I was hoping to purchase an RGBW one, looks like this is only RGB), I corrected my sketch to initialize the LED strip using this:

```
  strand = new Pixel_Strand(4,140);  // GPIO 4, 140 IC (controller) with 3 LED each = 420 LEDs on strip
```

Using Eve, selecting a red color gives me a full red strip.
Switching to blue, the strip remains red.
Switching back to red, the strip remains red.
Switching to neon green, the strip goes pink.
Switching to pink, the strip goes red.

If I simplify the code to control only the first pixel, I have realized that I need to "convert" the Hue of HSV to `180 - (px->H.getNewVal<float>()-180);`

And then, if I try to illuminate only the 2nd pixel (IC), then I get totally different colors.

So it is all over the place.
The WS2811 specs at https://cdn-shop.adafruit.com/datasheets/WS2812.pdf seems to say this is a 24-bit color.

Any tip to help me troubleshoot the issue? Where can I see raw bits sent to the strip?

**Comment by HomeSpan** [comment]
Let's start with a direct test of the Pixel code using the following script (change the PIN number accordingly):

```C++
#include "HomeSpan.h"

#define PIXEL_PIN 33    // set this to whatever pin you are using - note pin cannot be "input only"

Pixel testPixel(PIXEL_PIN);

void setup() {
 
  Serial.begin(115200);
  delay(1000);

  Serial.printf("Pixel Test on pin %d\n",PIXEL_PIN); 
}

//////////////////////////////////////

void loop(){
  Serial.print("Red\n");
  testPixel.set(Pixel::RGB(100,0,0));
  delay(2000);  

  Serial.print("Green\n");
  testPixel.set(Pixel::RGB(0,100,0));
  delay(2000);  
  
  Serial.print("Blue\n");
  testPixel.set(Pixel::RGB(0,0,100));
  delay(2000);  
}
```

This should cycle the color of the very first LED set on your strip from red to green to blue every two seconds.

Based on the above link, it looks like your strip needs 12 volts - make sure to power it directly from a 12V supply and connect the ground from the strip to the ground pin on the ESP32.   Also make sure the pin you choose to control the strip is connected to the digital INPUT on the strip (the strip allows for chaining, so one side will be a digital input and the other a digital output).

If this works for changing the colors of the first LED set on the strip, you can then modify the test to set all the LEDs on the set to the same color at the same time as a second test.

**Comment by frankonski** [comment]
Instead of blinking red, green, blue, it flashes red, blue green.

Connected to GPIO 4 of a [Keeyes dev board](https://www.amazon.ca/dp/B07QCP2451?ref=ppx_yo2ov_dt_b_product_details&th=1).
12 volts supply to the red wire, ground to white wire & ground pin of the dev board, and green wire to gpio pin 4.

**Comment by HomeSpan** [comment]
If it's blinking correctly but the colors do not match this may be a result of how the driver chip is connected to specific LEDs  within the strip.  I suppose different manufacturers can use the WS2811 differently.  I can add a hook that allows one to specify the order of the bytes. 

If you modify to control multiple pixels at the same time does it still operate correctly (apart from the green and blue being swapped)?

**Comment by frankonski** [comment]
> If you modify to control multiple pixels at the same time does it still operate correctly (apart from the green and blue being swapped)?

Yes it does. I tried with 140 pixels and the whole strip illuminates the same colors. Blue and green inversed.
I also tried with HSV, in case something was wonky with it as opposed to RGB, so I used with 0 and 359 (for red), 120 for green (got blue) and 240 for blue (got green).

**Comment by HomeSpan** [comment]
The fact that it works with 140 pixels is very good news - suggests that everything is working fine, with the exception of the color mapping.   I pulled the specs for the WS2811 and compared with the WS2812 and indeed the order of the colors when transmitting data to the chip is different!  I'll come up with a generic way of specifying the color order in the library so it can easily be configured to reflect the details of each chip and specific led strip.

Note the HSV functions is just a mathematical mapping to RGB, so if the RGB data is sent in the wrong order, the HSV won't work either.  I'll probably work on a patch this weekend --- will let you know when I post the changes to the dev branch.

**Comment by frankonski** [comment]
Thanks, and no rush. Illuminating a led strip is definitely less critical than having my alarm system send me push notifications! ;-)

**Comment by HomeSpan** [comment]
I added a new method to Pixel called `setColorMap(uint8_t *map)`.  It allows you to set the order in which the color bytes are transmitted.   Six color maps are pre-defined in the ColorMap namespace as follows (though I think only the first two will  ever be needed):

* ColorMap::GRB (this is the default)
* ColorMap::RGB (this is a common alternative)
* ColorMap::RBG
* ColorMap::BRG
* ColorMap::BGR
* ColorMap::GBR

It looks like only the first two are used by most manufacturers, though there is a lot of confusion as to which chips and devices use which order (for example, see this [discussion](https://forums.adafruit.com/viewtopic.php?p=846074)).

To use the new method, call it right after you instantiate a new Pixel.  For example, in the test sketch above, to switch the order of byte transmission to RGB (which I think is what you need for your specific device), you would add:

```C++
Pixel testPixel(PIXEL_PIN);
testPixel.setColorMap(ColorMap::RGB);   // add this new line
```

The new code is available in the "dev" branch (which now builds off of 1.9.0).

Let me know if this works.  If so I'll update the documentation accordingly.  I also have a nice test script that helps users determine what order they need, as well as whether or not they have an RGB or RGBW device.

**Comment by honear** [comment]
Just tested with the Sketch above and it works perfectly. May I ask, where would this new line go in the Pixels example?

**Comment by HomeSpan** [comment]
In the Pixel example you would enter it below the following line:

```C++
 pixel=new Pixel(pin);               // creates Pixel LED on specified pin <- Existing Line 79
 pixel.setColorMap(ColorMap::RGB);   // add this new line
```

I'm still contemplating the exact format of how this should work.  For now I've added `setColorMap` but I'm going to try out adding this as second parameter to the constructor itself, effectively replacing the *isRGBW* flag with a something like Pixel::Type::RGB, Pixel::Type::RGBW, Pixel::Type::GRB, etc.

**Comment by frankonski** [comment]
I will give it a try soon, but keep also in mind “all-white” led strips while re-working on your solution. I may end up having to purchase that instead, because Holidays colors all year long in the house will not pass the voting committee…

**Comment by frankonski** [comment]
I had to use `ColorMap::GBR`, but something is strange, setting to Red is now Pink.
```
  testPixel.setColorMap(ColorMap::BRG);   // r=g, b=r, g=b (Setting to Red gives me Green, ...)
  testPixel.setColorMap(ColorMap::RGB);   // r=b, b=g, g=r
  testPixel.setColorMap(ColorMap::GBR);   // r is pink!, g=g, b=b
```
.. unless I have an RGBW after all???

**Comment by HomeSpan** [comment]
If using RGB you see blue followed by green followed by red, use ColorMap::BGR

**Comment by frankonski** [comment]
ok.. definitely not RGBW, as when I set it, this is where the strip becomes multi-colored! :)

**Comment by frankonski** [comment]
> If using RGB you see blue followed by green followed by red, use ColorMap::BGR

No luck.

Red is green-ish
Green is red (a real red!)
Blue is blue.

**Comment by HomeSpan** [comment]
I'll give so more thought - these changes worked fine for @honear, but there may one something specific to your model.  Have you tried using a different driver (unrelated to HomeSpan) - perhaps the Adafruit NeoPixel library as a test?

**Comment by HomeSpan** [comment]
UPDATE:

The final version of the code is now checked-in on the "dev" branch.  To simplify use, I eliminated the recently-added `setColorMap()` method (which was mostly for initial testing purposes) and instead changed the constructor from `Pixel(int pin, boolean isRGBW);` to `Pixel(int pin, pixelType_t pixelType);` where *pixelType* should be set to one of the following:

* PixelType::RGB
* PixelType::RBG
* PixelType::BRG
* PixelType::BGR
* PixelType::GBR
* PixelType::GRB
* PixelType::RGBW
* PixelType::RBGW
* PixelType::BRGW
* PixelType::BGRW
* PixelType::GBRW
* PixelType::GRBW

Example: `Pixel testPixel(12, PixelType::RGBW);`

This allows you to simultaneously set the order of the RGB colors as well as specify whether your device also has a White LED.   In the new constructor, setting *pixelType* is optional.  If left blank, the default is PixelType::GRB, which matches the behavior in all previous versions of HomeSpan.  The original constructor, `Pixel(int pin, boolean isRGBW)` still works as expected, but I've flagged it as *deprecated* so you will receive a warning when compiling if used.

Finally, I added a standalone PixelTester sketch to help determine which setting to use for any given device.  You'll find it under Other Examples.

**Comment by frankonski** [comment]
Thanks, I ended up having to use `PixelType::GBR` for things to work.
Using your new sample that ends up increasing the brightness of R-G then B, I noticed that in the low brightness of Red, the strip was pink. Then as brightness increased, it went to bright red. 
So I took this, and re-uploaded the Holidays Light demo. With the demo and Eve app, I could now better control the colors, and using a bright red, within Eve with full brightness, the strip is red. Then decreasing the brightness from the app, the color is still red with lower brightness, and not pink. That code uses HSV, not RGB.

Nevertheless, I can now enjoy all but the Twinkle effect which has no lights at all and will try to figure out why and push a PR to update that different project if I can find the cause.

I owe you one for fixing this! Thanks!

and I'll let you close that issue if you are all done with it.

**Comment by HomeSpan** [comment]
Note that the Twinkle effect uses the white LED and requires an RGBW strip.

**Comment by frankonski** [comment]
Good point. I edited the effect to use the selected color, and now it all works fine.
I was also noticing flickering of incorrect colors, and I ended up using `setTiming(0.5, 2.0, 1.2, 1.3, 50);` to match the values found on page 3 of https://cdn-shop.adafruit.com/datasheets/WS2811.pdf. If I am not mistaken, this is especially meant for that, right? And no more flickering since then btw.

Is the documentation somewhere about the return value of the `update()` methods like at https://github.com/HomeSpan/HolidayLights/blob/main/HolidayLights.ino#L165 ?

Does this impact the speed of update? Speed of repeat?

**Comment by HomeSpan** [comment]
Yes, the `setTiming()` method is meant for exactly that -- really glad it solved the flickering issue.  I was doing some research and apparently some other Pixel controllers experienced the same problem. Certain manufacturers changed the timing cycle for their WS2811 chips a while back and flickering is the result of using timing parameters that are too close for the chip to distinguish a 1 from a 0.  One thing I noticed on the WS2811 doc is that it explicitly shows the order of bit transmission as RGB, but you needed to use GBR.  Perhaps Altilove connected the chip differently to the LEDs on their strip?  Regardless, as long as it works I guess all is good.

And yes, the return(20) in update() is the delay (in milliseconds) between updates to the effect.  The delay itself is implemented in the "alarm" code in lines 134-139.  Note that the update() method for the effects is not the same as the HomeSpan `update()`.  I probably should have picked a different name to avoid confusion.

**Comment by frankonski** [comment]
oh! I understand what you mean now. In `PixelStand.h`, I ended up:
* renaming `alarmTime` to `effectDelay`
* renaming the `SpecialEffect` function `virtual uint32_t update(){return(60000);}` to `virtual uint32_t run(){return(60000);}` everywhere where corrections were required, and now it all makes sense.

**Comment by HomeSpan** [comment]
Since everything seems like it is working with the new dev code I'll close out this issue.  Please feel free to re-open if needed.

**Comment by frankonski** [comment]
Hi, is it possible to reopen the issue? Something strange is happening. Still using the Holiday Lights project to get things to work, I realized that if I use the **Manual** effect, it initially works fine with the color selected. But then, if I toggle the color (using Eve app) back & forth between 7 preset colors, then all of a sudden the strip start flashing at ~1 second intervals.
Or, if I change the brightness multiple times, then the flashing also begins.
The only way I can get the flashing to stop is to keep changing the brightness and out of the blue the flashing stops and the color is solid and accurate.

If I set the effect on the Knight Rider, I can change the brightness & color as much as I want with no issues.

Setting to Random, the flashing will occur.

Setting to Twinkle (modified to use RGB & HSV), no issues.
Recetrack: no issues.

I got myself a full white strip that I will later try, but I would like to resolve this strange issue before.

**Comment by HomeSpan** [comment]
Unfortunately I won't be able to work on this for a while - and it may be very difficult to debug.  It may be that the speed of change is too rapid, in which case adding a short delay after the effect updates might help.  Other thing to check is wiring - make sure you connect the ground wire on both sides of the strip to the ESP32 ground.  

You can also try  to reduce the nPixels so that only the first half of the strip is used.  I'm not exactly sure what that would tell us, but it is a variable to test.

**Comment by frankonski** [comment]
Thanks, I tried something based on your suggestions.: I set the # of pixel to half of what the strip has: 70 (instead of 140).

Then I kinda noticed something and when on in isolating steps to reproduce:
* Using the Home app (I haven't confirmed yet with Eve), if I set the **brightness** to 73% and quickly change colors, no issues.
* If I increase brightness to 74% and go about changing colors, then it starts to flicker.

I will see if there is a specific count of color changes that the flicker starts to occur...

**Comment by frankonski** [comment]
update: no consistency in how many color changes I have to do before flickers happens... :(

**Comment by frankonski** [comment]
I also wired the other end of the ground, as suggested. No changes, flickering always starts at 74% brightness...

hum... I used an old router 12v power plug, says 1.0 amp output.
While the strip has 140 "pixels", this is actually 420 LEDs. I may be pulling to much amps.. need to look into this.

**Comment by frankonski** [comment]
With a multimeter in the loop, it got up to 1.7amps, but without flickering. If I remove the multimeter, it flickers. But the fact that I see 1.7 amps from an old Netgear power plug with max output saying 1.0 amp is a bad sign. I will find another power source and provide results later. ok to keep this issue closed for now.

**Comment by frankonski** [comment]
That was it. I found an old Netgear power plug, 12v 5amps output, and no flickering. All set!

**Comment by HomeSpan** [comment]
Excellent!  Thanks for the update.
